### PR TITLE
Update fusers.

### PR DESCRIPTION
**Pull request by sergeisakov**
Add detection of out of range qubits. Add clarifications on maximum qubit indices.

### CODE REVIEW

File: lib/fuser_basic.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ class BasicGateFuser final : public Fuser<IO, Gate> {
    * two-qubit gates will get fused. To respect specific time boundaries while
    * fusing gates, use the other version of this method below.
    * @param param Options for gate fusion.
-   * @param num_qubits The number of qubits acted on by 'gates'.
+   * @param max_qubit1 The maximum qubit index (plus one) acted on by 'gates'.
```
Comment (95-martin-orion) [inline lib/fuser_basic.h:59]: I assume this change is to more accurately describe situations where e.g. qubits {1, 2, 4} are being acted on? Would providing the exact set of qubits acted on be worthwhile, to ensure that a "skipped" qubit (3 in the previous example) isn't being erroneously acted on?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]

File: lib/fuser_basic.h [hunk 2]
```diff
@@ -360,17 +358,32 @@ class BasicGateFuser final : public Fuser<IO, Gate> {
...
-                               const GatesLat& gates_lat) {
-    for (unsigned q : qubits) {
-      if (!gates_lat[q].empty() && time <= gates_lat[q].back()->time) {
-        return true;
+  template <typename Gate2, typename GatesLat>
+  static bool ValidateGate(const Gate2& gate, unsigned max_qubit1,
+                           const GatesLat& gates_lat) {
+    for (unsigned q : gate.qubits) {
+      if (q >= max_qubit1) {
+        IO::errorf("gate qubits are out of range.\n");
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/fuser_basic.h]: Done.

File: lib/fuser_mqubit.h [hunk 3]
```diff
@@ -1052,17 +1050,34 @@ class MultiQubitGateFuser final : public Fuser<IO, Gate> {
...
-                               const GatesLat& gates_lat) {
-    for (unsigned q : qubits) {
-      if (gates_lat[q] != nullptr && time <= gates_lat[q]->val->parent->time) {
-        return true;
+  template <typename Gate2, typename GatesLat>
+  static bool ValidateGate(const Gate2& gate, unsigned max_qubit1,
+                           const GatesLat& gates_lat) {
+    for (unsigned q : gate.qubits) {
+      if (q >= max_qubit1) {
+        IO::errorf("gate qubits are out of range.\n");
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/fuser_mqubit.h]: Done.

File: lib/fuser_basic.h (hunk 1 above)
Comment (sergeisakov) [reply lib/fuser_basic.h:59]: This change is to describe situations where e.g. qubits {1, 2, 4} and max_qubit1 (previously num_qubits) is smaller than 4+1. max_qubit1 is used to allocate an array in line 165 and there will be out of bound array access if any qubit acted on by gates is greater or equal to max_qubit1. It is OK if there are "skipped" qubits.

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**